### PR TITLE
✨ Migrate to new Sentry API

### PR DESCRIPTION
**Pull request by greenfrogs**
## What does this PR do?
Migrates from Raven to the new Sentry API for Django and React. Part of #3053

## ✅ Pull Request checklist
- [x] Initial migration
- [x] Scrub personal data
- [ ] Gencache errors
- [ ] Add release information to Sentry (wait for #3114)

## 🚨 Is this a breaking change for API clients?
No

## :squirrel: Deploy notes
Update new DSN keys for Django in .env and add new SENTRY_DSN_REACT in Travis for webpack.

### CODE REVIEW

File: backend/uclapi/requirements.txt
Comment (Feverfew) [inline backend/uclapi/requirements.txt:30]:
```suggestion
sentry-sdk==0.20.1
```
Benefits of not reviewing... although test that it doesn't break anything :P

File: backend/uclapi/uclapi/ourcelery.py
Comment (Feverfew) [inline backend/uclapi/uclapi/ourcelery.py:24]: We need the `before_send` keyword argument to be filled as well: https://docs.sentry.io/platforms/python/configuration/filtering/#using-beforesend

This is particularly important for us as we have the legacy of putting confidential information as a query parameter, which we should prevent from being stored on other servers. So make sure we strip tokens, client_id/state etc.

Also strip anything else that we don't need for debugging but is sensitive.

File: backend/uclapi/uclapi/settings.py
Comment (Feverfew) [inline backend/uclapi/uclapi/settings.py:164]: does this work for us? We don't use the default Django authentication stuff.

File: frontend/src/pages/Warning.jsx
Comment (Feverfew) [inline frontend/src/pages/Warning.jsx:1]: Anyway that this boilerplate can be avoided.

**Review by Feverfew — CHANGES_REQUESTED** [review]
Fairly straightforward. Just a few questions I'd like answered.

Await review from Huey in terms of frontend and Zak so that he puts in the required stuff in the `.env`.

File: backend/uclapi/uclapi/settings.py
Comment (greenfrogs) [reply backend/uclapi/uclapi/settings.py]: We don't get the Django auth benefit however this sends us additional information about the user client (such as ip address and user agent). In the interest of privacy I have disabled it.

File: frontend/src/pages/Warning.jsx
Comment (greenfrogs) [reply frontend/src/pages/Warning.jsx:1]: This was the shortest I could make it, unless @hueyy has any ideas. From my understand each page is an entry point so we need it on each of them.

File: backend/uclapi/uclapi/ourcelery.py
Comment (greenfrogs) [reply backend/uclapi/uclapi/ourcelery.py:24]: It turns out that sentry sends quite a bit more information than I was expecting, some in several different forms (I counted the URL with get parameters provided in at least 5 different nodes of the tree). Sentry has some server side scrubbing but ideally we would not send anything sensitive.

This is made harder by the event tree being very variable and has resulted in this recursive explorer checking for dict/list entries. We then check against a list of known keys which contain sensitive information and remove those entries. For the URL entries and some trace backs (the get parameters are passed in as arguments so sometimes shown in the stack trace), this approach uses regex's to match the possibly generated tokens, client ids, and client secrets. I am not sure of a better solution (without regex) which won't redact too much debugging information.

File: frontend/src/pages/Warning.jsx
Comment (hueyy) [reply frontend/src/pages/Warning.jsx:1]: Yeah the way it's set up right now requires this. Not worth doing something fancy like injecting it into the React dependency via webpack imo.

**Review by hueyy — APPROVED** [review]
LGTM!

File: backend/uclapi/uclapi/settings.py
Comment (Feverfew) [inline backend/uclapi/uclapi/settings.py:164]: Lets add in `code` and `client_id` for good measure.

**Review by Feverfew — APPROVED** [review]
Otherwise LGTM, Let Zak merge so that he updates the `.env`.

**Comment by greenfrogs** [comment]
Please do not merge until I have added the gencache errors and added deployment to GitHub Actions.

**Comment by Feverfew** [comment]
> Please do not merge until I have added the gencache errors and added deployment to GitHub Actions.

Marked as draft :)

**Review by zipy124 — APPROVED** [review]

**Pull request merged**
